tpm_futurepcr/LogEvent.py

```diff
@@ -79,33 +103,96 @@ def show(self):
+        device_path_vec = parse_efi_device_path(binary_reader.read(device_path_len))
+        self.data = _EFIBSAData(image_location, image_length, image_lt_address, device_path_vec)
+
+        unix_path = self._device_path_to_unix_path()
+        if unix_path is None:
+            # This might be a firmware item such as the boot menu.
+            if not self.allow_unexpected_bsa:
+                logger.error("Unexpected boot events found. Binding to these PCR values "
+                             "is not advised, as it might be difficult to reproduce this state "
+                             "later. Exiting.")
+                exit(1)
+            logger.warning("couldn't map EfiBootServicesApplication event to a Linux path")
+        self.unix_path = self.substitute_bsa_unix_path.get(unix_path, unix_path)
+
+    def _device_path_to_unix_path(self) -> Path | None:
+        dir_path = None
+        unix_path = None
+        for pp in self.data.device_path_vec:
+            if pp.type == DevicePathType.MediaDevice:
+                if pp.subtype == MediaDevicePathSubtype.HardDrive:
+                    dir_path = find_mountpoint_by_partuuid(pp.part_uuid)
+                    if not dir_path:
+                        raise Exception("could not find mountpoint for partuuid %r" % pp.part_uuid)
+                if pp.subtype == MediaDevicePathSubtype.FilePath:
+                    unix_path = dir_path / Path(pp.file_path)
+            if pp.type == DevicePathType.End:
+                break
+        return unix_path
+
+    def show(self):
+        if logger.level == logging.DEBUG:
+            for i in hexdump(self.data):
+                logger.debug(i)
+            ed = dataclasses.asdict(self.data)
+            logger.debug(pformat(ed))
         else:
-            self.data = binary_reader.read(event_size)
+            logger.verbose("Path vector:")
+            for p in self.data.device_path_vec:
+                type_name = getattr(p["type"], "name", str(p["type"]))
+                subtype_name = getattr(p["subtype"], "name", str(p["subtype"]))
+                file_path = p.get("file_path", p["data"])
+                logger.verbose("  * %-20s %-20s %s", type_name, subtype_name, file_path)
+
+    def _hash_pecoff(self, hash_alg="sha1"):
+        try:
+            with open(self.unix_path, "rb") as fh:
+                fpr = signify.fingerprinter.AuthenticodeFingerprinter(fh)
+                fpr.add_authenticode_hashers(getattr(hashlib, hash_alg))
+                return fpr.hash()[hash_alg]
+        except FileNotFoundError:
+            logger.info("File %s could not be opened. Continuing with the log-provided extend value", self.unix_path)
+            raise ValueError
+
+    def next_extend_value(self, current_extend_value: bytes, hash_alg: str = "sha1") -> bytes:
+        return self._hash_pecoff(hash_alg)
+
+
+def logEventFactory(binary_reader, tpm_version, tcg_hdr, substitute_bsa_unix_path, allow_unexpected_bsa):
```

Comment:
A minor thing, but I'd use CamelCase if it's supposed to masquerade as a class constructor, standard snake_case for normal functions – not Java style.

Reply:
Agreed, and done.